src/irmin-pack/IO.ml

```diff
@@ -232,25 +259,27 @@ module Unix : S = struct
     mkdir (Filename.dirname file);
     match Sys.file_exists file with
     | false ->
-        let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
-        let raw = Raw.v x in
-        Raw.unsafe_set_offset raw 0L;
-        Raw.unsafe_set_version raw current_version;
-        v ~offset:0L ~version:current_version raw
+        if readonly then v ~offset:0L ~version:current_version None
+        else
+          let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+          let raw = Raw.v x in
+          Raw.unsafe_set_offset raw 0L;
+          Raw.unsafe_set_version raw current_version;
+          v ~offset:0L ~version:current_version (Some raw)
```

Comment:
Totally subjective, but I'd prefer pushing the `if` further into the logic:

```ocaml
let raw_opt =
  match readonly with
  | true -> None
  | false ->
      let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
      let raw = Raw.v x in
      Raw.unsafe_set_offset raw 0L;
      Raw.unsafe_set_version raw current_version;
      Some raw
in
v ~offset:0L ~version:current_version raw_opt

Reply:
There's also a small TOCTOU race condition here (which was already present): the file may be created and filled with data after line 260 but before line 264.

I think we should pass the `O_EXCL` flag and drop into the `true` case if it raises a "File already exists" exception.

Reply:
> There's also a small TOCTOU race condition here (which was already present): the file may be created and filled with data after line 260 but before line 264.

Good catch. I think the race can only occur when two rw instances (of the same process) try to open a "fresh" pack concurrently. In this case, it should be ok, they write the same thing.
That's the only case I see, because as soon as one rw instance opened a pack, it is then cached for future instances.

Reply:
Do they always write the same thing? The two RW instances might pass different `~version`s.

Reply:
`version` is the `current_version` which is hardcoded in `pack.ml`. And I think it will stay this way but I'm not really sure, @pascutto, @samoht ?

Reply:
It can change if we decide to change the format of irmin-pack.

Reply:
But we haven't implemented yet the logics to handle such cases : we need to do something if the version on disk is different than the version we expect ; a simple solution will just be to fail with an informative error message -- a more complex solution would be to upgrade the storage to the new on-disk format.

Reply:
In that case, they will always pass the same thing :slightly_smiling_face: If we require all constructed instances to have the same version, I suggest that we do one of:
 - detail this constraint in a doc-comment of `IO.S.v` (and comment here that the constraint is being relied upon)
 - move `Irmin_pack.current_version` to a central `Version` module avoid passing it in at all

Reply:
Not if you run 2 different instance of irmin-pack concurrently (with different on-disk format)

Reply:
Good point; I forgot this was a cross-process race too. In that case I revert to my earlier suggestion of a defensive `O_EXCL` flag.

Reply:
I added the `O_EXCL` flag here (and removed it from when we [open the file and the file exists](https://github.com/mirage/irmin/pull/955/files#diff-1a4152c6731a5518a2728ab388cf55d2b58315685a58019c5f0eb8f30ce6b01dR245)).